### PR TITLE
[POC] Replaced document-manager by event-sourcing and CQRS

### PR DESCRIPTION
**Pull request by wachterjohannes**
THIS IS AN EXPERIMENT AND FAR AWAY FROM BEEING PERFECT.

| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | none
| License | MIT

#### What's in this PR?

This PR wraps the document-manager functionality into a projection for CQRS and Event-Sourcing.

#### Why?

This could be a replacement for the document-manager. All the functionality which is covered by the document-manager will be moved to an `AggregateRoot`.

#### Database Schema

```sql
DROP TABLE IF EXISTS `event_streams`;
CREATE TABLE `event_streams` (
  `no` bigint(20) NOT NULL AUTO_INCREMENT,
  `real_stream_name` varchar(150) COLLATE utf8_bin NOT NULL,
  `stream_name` char(41) COLLATE utf8_bin NOT NULL,
  `metadata` json DEFAULT NULL,
  `category` varchar(150) COLLATE utf8_bin DEFAULT NULL,
  PRIMARY KEY (`no`),
  UNIQUE KEY `ix_rsn` (`real_stream_name`),
  KEY `ix_cat` (`category`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
 
DROP TABLE IF EXISTS `projections`;
CREATE TABLE `projections` (
  `no` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(150) COLLATE utf8_bin NOT NULL,
  `position` json DEFAULT NULL,
  `state` json DEFAULT NULL,
  `status` varchar(28) COLLATE utf8_bin NOT NULL,
  `locked_until` char(26) COLLATE utf8_bin DEFAULT NULL,
  PRIMARY KEY (`no`),
  UNIQUE KEY `ix_name` (`name`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
 
DROP TABLE IF EXISTS `snapshots`;
CREATE TABLE `snapshots` (
  `aggregate_id` varchar(150) COLLATE utf8_bin NOT NULL,
  `aggregate_type` varchar(150) COLLATE utf8_bin NOT NULL,
  `last_version` int(11) NOT NULL,
  `created_at` char(26) COLLATE utf8_bin NOT NULL,
  `aggregate_root` blob,
  UNIQUE KEY `ix_aggregate_id` (`aggregate_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
```

#### To Do

- [ ] Article-Pages

### CODE REVIEW

File: Controller/ArticleController.php [hunk 1]
```diff
@@ -322,9 +357,20 @@ public function putAction(Request $request, $uuid)
 
         $this->get('sulu_hash.request_hash_checker')->checkHash($request, $document, $document->getUuid());
 
-        $this->persistDocument($data, $document, $locale);
-        $this->handleActionParameter($action, $document, $locale);
-        $this->getDocumentManager()->flush();
+        $this->getCommandBus()->dispatch(
+            new UpdateArticle(
```
Comment (chirimoya) [inline Controller/ArticleController.php]: `ModifyArticle`

File: Controller/ArticleController.php [hunk 2]
```diff
@@ -604,4 +613,45 @@ protected function getMetadataFactory()
...
+            new CreateArticleHandler($userRepository, $this->get('sulu_content.structure.factory'))
+        );
+        $router->route(PublishArticle::class)->to(new PublishArticleHandler($userRepository));
+        $router->route(UnpublishArticle::class)->to(new UnpublishArticleHandler($userRepository));
+        $router->route(UpdateArticle::class)->to(
+            new UpdateArticleHandler($userRepository, $this->get('sulu_content.structure.factory'))
+        );
+        $router->attachToMessageBus($commandBus);
+
+        $userProjector = new ArticleDocumentProjector(
```
Comment (chirimoya) [inline Controller/ArticleController.php]: `$articleDocumentProjector`

File: Builder/ArticleIndexBuilder.php [hunk 3]
```diff
@@ -69,6 +69,6 @@ private function buildForManager(Manager $manager, $destroy)
         }
 
         $this->output->writeln(sprintf('Drop and create index for "<comment>%s</comment>" manager.', $name));
-        $manager->dropAndCreateIndex();
+        // $manager->dropAndCreateIndex();
```
Comment (wachterjohannes) [inline Builder/ArticleIndexBuilder.php]: to be removed

**Comment by niklasnatter** [comment]
Replaced by #553

**Pull request closed without merging**